Add LCDIF support for RT700

KATE-WANG-NXP opened the pull request:
1.Update nxp_dcnano_lcdif display driver
2.Add new mipi_dbi driver nxp_dcnano_lcdif_dbi
3.Add new display driver for panels which have bl pin and use DCS commands

zejiang0jason commented:
Hi @KATE-WANG-NXP , please pay attention to the code format, such as: `Tab` instead of `space`, file end `newline`. Thanks.

danieldegrasse commented on drivers/display/Kconfig.mcux_dcnano_lcdif:
```diff
@@ -58,4 +58,4 @@ config MCUX_DCNANO_LCDIF_EXTERNAL_FB_ADDR
 endif # MCUX_DCNANO_LCDIF_EXTERNAL_FB_MEM
 
-endif # DISPLAY_MCUX_ELCDIF
+endif # DISPLAY_MCUX_LCDIF
```
```suggestion
endif #DISPLAY_MCUX_DCNANO_LCDIF
```

Should match the Kconfig name used on line 13 in this file

KATE-WANG-NXP replied:
Thanks, fixed.

danieldegrasse commented on drivers/display/display_mcux_dcnano_lcdif.c:
```diff
@@ -278,6 +297,29 @@ static DEVICE_API(display, mcux_dcnano_lcdif_api) = {
 #endif
 
+#if DT_ENUM_IDX_OR(DT_NODELABEL(lcdif), version, 0) == 1
+#define MCUX_DCNANO_LCDIF_FB_CONFIG(n)					\
```
Nit here- please align the `\` backslashes on each line, to remain stylistically consistent with the rest of the file.

KATE-WANG-NXP replied:
Thanks, fixed.

danieldegrasse commented on drivers/display/display_mcux_dcnano_lcdif.c:
```diff
@@ -290,14 +332,10 @@ static DEVICE_API(display, mcux_dcnano_lcdif_api) = {
-		.next_idx = 0,							\
-		.pixel_bytes = MCUX_DCNANO_LCDIF_PIXEL_BYTES(n),		\
-	};									\
+		MCUX_DCNANO_LCDIF_FB_CONFIG(n)									\
```
Same request here- using `git diff` can often make it easier to spot this alignment issue with the `\`, if your editor isn't helping.

danieldegrasse commented on dts/bindings/display/nxp,dcnano-lcdif.yaml:
```diff
@@ -32,3 +32,11 @@ properties:
     description:
       LCD data bus width. The default is set to the reset value of 24-bit
+
+  version:
```
Not opposed to this approach, but is this an IP _revision_, or a new IP block? If it is a new IP block, we can actually support multiple compatibles within one driver (and we should, in that case). See here for an example: https://github.com/zephyrproject-rtos/zephyr/blob/e60f04096cd8d9227c8279ca830e72bc1b5b8b1c/drivers/i2c/i2c_tca954x.c#L204

KATE-WANG-NXP replied:
Thanks, I did not know about this usage before. I think it is an IP revision, for DC8000 there are extra registers to set for a frame configuration compared to DCNano. Actually I did not find any similar use case of adding a new binding for IP revision so I'm not sure it is good practice. I did find using kconfig symbols to isolate extra register configurations but I think it is obsolete? https://github.com/zephyrproject-rtos/zephyr/blob/e60f04096cd8d9227c8279ca830e72bc1b5b8b1c/drivers/counter/counter_mcux_snvs.c#L279

KATE-WANG-NXP replied:
Yes DC8000 has a lot more registers compared with DCNano, because in DC8000 apart form the cursor layer, there are 3 graphical layers instead of 1, to perform color key, or porter duff etc. Since these functions are not used in Zephyr, I was wondering that it would be unnecessary and confusing to add property for each one, so I just added a version binding, similarly treated in MCUXpresso SDK lcdif driver.

danieldegrasse replied:
Yeah, I think the version binding is probably the right choice- we can split this into a new binding in the future if we end up utilizing the porter-duff blending or similar features

danieldegrasse commented on drivers/mipi_dbi/Kconfig.nxp_dcnano_lcdif:
```diff
@@ -0,0 +1,10 @@
+	bool "MIPI DBI driver for NXP DCnano LCDIF"
+	default y
+	depends on DT_HAS_NXP_DCNANO_LCDIF_DBI_ENABLED
+    select PINCTRL
```
Nit- fix indentation

danieldegrasse commented on drivers/mipi_dbi/mipi_dbi_nxp_dcnano_lcdif.c:
```diff
@@ -0,0 +1,430 @@
+	if (0 != (status & kLCDIF_Display0FrameDoneInterrupt))
+	{
+		/* Handle multi chunk with mipi. */
+#if CONFIG_MIPI_DSI
```
I'm unclear why we have dependencies on `CONFIG_MIPI_DSI` here?

From what I understand about the DC8000 peripheral, it can support MIPI DBI (8080 and 6800), and it can support MIPI DPI, which would be sent via a DSI phy to the display.

A few questions here:
- The MIPI DPI support would be used for displays that operate in video mode (IE no graphics RAM, require constant refresh). An example would be the RK055HDMIPI4MA0 display
- The DSI PHY on this SOC supports sending MIPI DSI commands directly, right? This would be used for displays like the G1120B0MIPI, correct? Or can the DC8000 still drive those displays?

If I'm understanding the peripheral right, then we probably need something like the following:
- A new compatible for the peripheral- this is clearly different from the DCNANO IP present on the RT500
- A MIPI-DBI driver that supports driving displays in 6800 or 8080 mode
- A MIPI-DSI driver for the MIPI-DPHY (if this IP is reused from the RT500, we can reuse the existing driver)
- Additions to the DCNANO LCDIF driver to support this new DC8000 IP (using the trick I mentioned for multiple compatibles)

Critically, the DC8000 should only have *one* compatible string. Users can select which driver gets used with the IP based on which Kconfigs they enable. We get into a bit of a strange situation though, since it seems this IP can be used in MIPI-DBI mode, or as a display driver (enabled with `CONFIG_DISPLAY`). We probably need to make it so that `CONFIG_DISPLAY_MCUX_DCNANO_LCDIF` depends on `CONFIG_MIPI_DBI` being disabled.

KATE-WANG-NXP replied:
_The DSI PHY on this SOC supports sending MIPI DSI commands directly, right? This would be used for displays like the G1120B0MIPI, correct? Or can the DC8000 still drive those displays?_
Yes, on RT700 the panel G1120B0MIPI is driven by DC8000 and MIPI DSI combined. We can also use the RT500 legacy way to drive G1120B0MIPI by sending MIPI DSI commands and data directly on APB, but since RT700 does not have smartdma, this way can be rather slow, and does not support pixel format convertion, eg. the frame buffer in memory is stored in RGB888 format while the data sent on bus in D16RGB565 format as G1120B0MIPI required. To use both DC8000 and the MIPI DSI to drive the panel, there are many additional actions to do on the MIPI DSI side. Since the internal interface between these 2 modules have limited buffering, if the data to be sent is too large, the driver will have to chop up the data and sent one piece at a time. So for G1120B0MIPI the CONFIG_MIPI_DSI has to be enabled, for displays in 6800 or 8080 mode, there is no need to enable CONFIG_MIPI_DSI.
_A new compatible for the peripheral- this is clearly different from the DCNANO IP present on the RT500_
Yes I created a new compatible nxp,dcnano_lcdif_dbi for this driver drivers/mipi_dbi/mipi_dbi_nxp_dcnano_lcdif.c, take reference from the compatible nxp,dcnano_lcdif of drivers/display/display_mcux_dcnano_lcdif.c(not sure if it is a good practice since they belong to different driver groups). They all use the same IP DC8000 on RT700, but as you pointed out DC8000 can either be in DBI or in DPI mode so only one compatible at the same time, so in my test example, I change the compatible in the display's overlay(still not sure if it is a good practice since I cannot find similar cases).
https://github.com/zephyrproject-rtos/zephyr/commit/476b7bb1df6786620fd759ce2b6e9ab25703cc23#diff-1b9a0139b8cc287d78c4bb150fc04ccc161c82f8fea192fe9aa1ecc9a177c18aR18
_We probably need to make it so that CONFIG_DISPLAY_MCUX_DCNANO_LCDIF depends on CONFIG_MIPI_DBI being disabled._
Thanks, will update the dependency.

danieldegrasse replied:
To add to the above, MIPI DBI (6800/8080) mode should be handled by the MIPI DBI driver, and MIPI-DSI should be handled by the MIPI-DSI driver.

Seeing this peripheral now, I am wondering if we should be merging the MIPI-DBI and MIPI-DSI driver classes at some point, since they do use the same command set (MIPI DCS). That is a broader question though

KATE-WANG-NXP replied:
Ok, so the DC8000 is sending video data to the MIPI DSI, which then sends it to the RM67162 controller on the G1120B0MIPI? _**Yes.**_ 
Do we have the ability to perform partial writes using the DC8000? I ask because the RM67162 controller has internal GRAM, so we should be able to do partial refreshes (whereas a standard video mode MIPI DSI controller would not accept partial refreshes) **_Yes. We can use DC8000 to chose which part of the area on screen to update by setting the coordinates._**
I'm still not clear where MIPI DBI fits in this chain. For the RM67162, I would expect we would perform partial refreshes via the MIPI DSI API- 
**_I would prefer this way too so RM67162 driver can still be used as the display driver same as on RT500, but when using the DC8000 and MIPI DSI combined to drive the panel, sending command/data is controlled by DC8000 instead of MIPI-DSI, MIPI-DSI is simply performing like a bridge. So in order to use DC8000 to send DCS command, I added nxp_dcnano_lcdif_dbi driver._**
If we need to integrate the DC8000 into this chain, the best way to do that is likely to use the DC8000 as the Zephyr display driver, and have the RM67162 driver only handle the init phase of the display setup. This is what we do for video mode displays currently. **Yes that is exactly what I'm dong for RT700 display example for G1120B0MIPI, refering to the HX8394 on RT500, using nxp_dcnano_lcdif_dbi as display driver, only leaving the init to RM67162 driver**.

To add to the above, MIPI DBI (6800/8080) mode should be handled by the MIPI DBI driver, and MIPI-DSI should be handled by the MIPI-DSI driver. 
Seeing this peripheral now, I am wondering if we should be merging the MIPI-DBI and MIPI-DSI driver classes at some point, since they do use the same command set (MIPI DCS). That is a broader question though. **_I was thinking about it at some point too, but since the hardware interfaces of MIPI-DBI and MIPI-DSI are different similar to parallel and serial ports, plus MIPI-DSI has DPHY to consider, I would prefer leaving them seperated._**

KATE-WANG-NXP replied:
I think these changes are all included in this PR?
**Any additional configuration needed for the DC8000 versus the DCNANO**
You mean the update DPI mode right? It's in https://github.com/zephyrproject-rtos/zephyr/pull/82635/commits/ede2ef3931e497de325b8cd56bcc1eeeb1930ce8
**Changes to the display_write function to support partial area updates when using the DC8000**
I believe it shall be done by the DBI driver? Or you mean the lcdif DCNano driver shall support both DPI or DBI mode?
**A new MIPI-DBI driver for the DC8000. **
It's in https://github.com/zephyrproject-rtos/zephyr/pull/82635/commits/c2f74b8f1e7f7f111bb57e4d4a9e7a610efb303f
**Any changes needed to the MIPI DSI driver to enable it to act as a bridge for the DCNANO (may not need any, because we do this on the RT500 already for video mode)**
Its also added in https://github.com/zephyrproject-rtos/zephyr/pull/82635/commits/c2f74b8f1e7f7f111bb57e4d4a9e7a610efb303f protected by CONFIG_MIPI_DSI macro. I think it is not appropriate to add it in the MIPI DSI driver since the operations are all very low level.

danieldegrasse replied:
> I think these changes are all included in this PR?
Any additional configuration needed for the DC8000 versus the DCNANO
You mean the update DPI mode right? It's in https://github.com/zephyrproject-rtos/zephyr/commit/ede2ef3931e497de325b8cd56bcc1eeeb1930ce8

Yup, this looks good- keep in mind I'm not deeply familiar with this hardware, but I think this is the right approach.

> Changes to the display_write function to support partial area updates when using the DC8000
I believe it shall be done by the DBI driver? Or you mean the lcdif DCNano driver shall support both DPI or DBI mode?

I think the lcdif DCNano driver should support video mode and command mode via the MIPI-DSI link. Essentially, I'd expect the `display_write` implementation in the LCDIF DCNANO driver to support partial writes if using a command mode MIPI DSI display

> A new MIPI-DBI driver for the DC8000.
It's in https://github.com/zephyrproject-rtos/zephyr/commit/c2f74b8f1e7f7f111bb57e4d4a9e7a610efb303f
Any changes needed to the MIPI DSI driver to enable it to act as a bridge for the DCNANO (may not need any, because we do this on the RT500 already for video mode)
Its also added in https://github.com/zephyrproject-rtos/zephyr/commit/c2f74b8f1e7f7f111bb57e4d4a9e7a610efb303f protected by CONFIG_MIPI_DSI macro. I think it is not appropriate to add it in the MIPI DSI driver since the operations are all very low level.

I think the changes should be in the MIPI DSI driver- the DBI driver will only be use when we we are running the DCNANO with a display controller using 8080/6800 mode, and the DSI driver will be used when we use a display controller in MIPI DSI video/command mode, right?

KATE-WANG-NXP replied:
When MIPI-DSI is used as the bridge, what should be done on the MIPI-DSI side is different for command mode from video mode. In video mode, once the MIPI-DSI is initialized, there is nothing more needed for frame update, but in command mode, for each write the DSI buffer/payload/send level has to be reconfigured. So if the lcdif DCNano driver is going to support the command mode, it just has to call those low level SDK driver APIs, like currently my new DCNANO MIPI-DBI driver does. And I think you are right, the DCNANO MIPI-DBI driver shall only handle 8080/6800 mode, and keep the low level MIPI-DSI operation all inside the MIPI-DSI driver. So I'm thinking to support DCNano command mode via the MIPI-DSI link, is it better to update the MIPI-DSI driver, let user to configure whether to use the DCNano command mode for acceleration, since for RT700 the G1120B0MIPI can still only use the MIPI-DSI despite the low speed. If DCNano command mode is used, the MIPI-DSI driver will call DCNANO MIPI-DBI APIs then those low level SDK driver APIs, instead of writing data to DSI APB buffer. In this case the RM67162 driver can still be used as the display driver for G1120B0MIPI.

danieldegrasse commented on drivers/display/display_mipi_dbi.c:
Do displays actually exist that only require the DCS command set in order to initialize and display data? Every display I've ever encountered does use the DCS command set for writing to graphics RAM, but also requires a long set of vendor specific commands to set up support for the display properly

KATE-WANG-NXP replied:
Yes, to drive the panel G1120B0MIPI using both DC8000 and MIPI DSI, the driver display_rm67162.c can not be used, so I created this driver based on display_rm67162.c. I just updated the write and set_pixel_format by using mipi_dbi_write_display to send the data instead of mipi_dsi_dcs_write. Then set CONFIG_DISPLAY_MIPI_DBI=y and change the chosen zephyr display to display_mipi_dbi: https://github.com/zephyrproject-rtos/zephyr/commit/476b7bb1df6786620fd759ce2b6e9ab25703cc23#diff-1b9a0139b8cc287d78c4bb150fc04ccc161c82f8fea192fe9aa1ecc9a177c18aR12
The panel initilization function still is rm67162_init. This driver can also be used for panels similar to G1120B0MIPI, for example ZC143AC72MIPI, this panel is a replacement of G1120B0MIPI since it is discontinued.

danieldegrasse replied:
So what I'm confused on here is why we are using the DBI driver here under the hood at all. The RM67162 uses MIPI-DSI. Why can't the display_rm67162 driver be used on the RT700? Is is because the DC8000 is used to drive the display data? In that case (as I said in another comment) the existing RM67162 driver should be used for display init, then the actual display driver (which has display_write called on it) should implement support for the DC8000.

KATE-WANG-NXP replied:
Sorry I was on Chinese new year leave and missed your comment. 
> The RM67162 uses MIPI-DSI. Why can't the display_rm67162 driver be used on the RT700?
Because RT700 does not have smartdma, so MIPI-DSI can only send commands and data through writing APB buffer word by word using CPU, this way is too slow to meet most application requirement, details are as mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/82635#discussion_r1879554433.
But the MIPI-DSI can use another acceleration way similar to involving the smartdma, that is to use the DCNano DBI mode to send data to MIPI-DSI APB from memory, once the buffer address, pixel type etc. are configured in DCNano. Comparing to CPU writing the APB register, this way is much efficient. This way is a little like driving the RK055MHD091A0 using nxp,dcnano-lcdif through MIPI-DSI interface, but on MIPI-DSI side it is very different, for RK055MHD091A0 all it has to be done is call mipi_dsi_attach in hx8394_init, while for MIPI-DSI to use DCNano DBI acceleration, each memory write requires many operations of the low level mcux mipi-dsi driver.
In this case here is my latest proposal, to drive G1120B0MIPI, first enable DCNano DBI driver 1ebfa11317fd5218380b1c6ec41850bdc7c8f37d, this driver is to drive 8080/6800 panel when used alone. (Although RT700 does not support this kind of panel yet).
Second, update dsi_mcux_2l to use DCNano DBI driver for memory write for acceleration on RT700 abd0e7c9e6c07d8a70b090b8d26af3ec4ccbbd7b.

KATE-WANG-NXP committed:
drivers: display: Fix typo in kconfig comment of nxp,dcnano-lcdif.

Should be DISPLAY_MCUX_DCNANO_LCDIF instead of DISPLAY_MCUX_ELCDIF.

Signed-off-by: Kate Wang <yumeng.wang@nxp.com>

KATE-WANG-NXP committed:
drivers: display: Update nxp,dcnano-lcdif to support IP change on RT700

Update nxp,dcnano-lcdif to support IP change on RT700. There are extra
registers need to be configured for the lcdif on RT700. Add new binding
item "version" to tell which version of the IP the SoC has.

Signed-off-by: Kate Wang <yumeng.wang@nxp.com>

KATE-WANG-NXP committed:
drivers: clock_control: update clock_control_mcux_syscon driver for RT700

Update pixel clock control to support RT700.

Signed-off-by: Kate Wang <yumeng.wang@nxp.com>

rruuaanng commented on dts/bindings/mipi-dbi/nxp,mipi-dbi-dcnano-lcdif.yaml:
```diff
@@ -0,0 +1,137 @@
+  DBI settings for NXP DCnano LCD controller. DCNano is used in DPI mode by default,
+  when used in DBI mode the compatible need to change in example overlay. For example:
+
+  &lcdif {
```
Hmmm... I'm wondering if the example provided here is a bit redundant, is this a rare behavior? If not, I think it doesn't need to be over-explained.
(This makes the first half of the file look a bit unattractive.)

KATE-WANG-NXP replied:
I think it is quite uncommon, I do not know other similar use case myself.

KATE-WANG-NXP commented:
Hi all, please help review this PR, the v4.1.0(Feb14) is quite close, thanks

KATE-WANG-NXP commented:
Rebased main branch for https://github.com/zephyrproject-rtos/zephyr/pull/82959

dleach02 commented:
@danieldegrasse you engaged the most on this PR with @KATE-WANG-NXP. Are you good with the the current state so we can get this merged for v4.1

kartben commented:
@dleach02 @KATE-WANG-NXP the PR needs a rebase though

KATE-WANG-NXP commented:
Hi all, please help review/merge this PR https://github.com/zephyrproject-rtos/hal_nxp/pull/507, so that I can rebase to resolve conflict and replace the sync with real commit. Thanks

dleach02 commented:
I'll handle the rebase and merging HAL

dleach02 commented:
@kartben I merged the HAL and rebased this PR for @KATE-WANG-NXP

zejiang0jason commented on boards/nxp/mimxrt700_evk/mimxrt700_evk_mimxrt798s_cm33_cpu0.dts:
```diff
@@ -107,6 +143,14 @@
 
+&gpio2 {
+	status = "okay";
+};
```
seems not used.

mmahadevan108 replied:
Please delete this node if not used.

dleach02 commented:
another PR merged and forced another rebase...

danieldegrasse commented on drivers/mipi_dsi/dsi_mcux_2l.c:
```diff
@@ -560,6 +581,83 @@ static ssize_t dsi_mcux_transfer(const struct device *dev, uint8_t channel,
+			desc.first_write = false;
+		}
+
+		mipi_dbi_write_display(config->mipi_dbi, NULL, msg->tx_buf,
```
Why are we using the `mipi_dbi` API here within the MIPI DSI driver? The drivers should be independent of one another. Ideally, the user should be able to use this IP with drivers in two modes:
- when using a DSI display, use the LCDIF driver paired with the DSI_MCUX_2L driver
- when using a DBI display, use the display controller specific driver, paired with the lcdic_dbi driver

DBI mode (IE 8080,6800, and SPI) should be independent from DSI mode (video and command mode)

KATE-WANG-NXP replied:
Please see my earlier comment:
https://github.com/zephyrproject-rtos/zephyr/pull/82635#discussion_r1942502499

KATE-WANG-NXP replied:
Yes this is very accurate.
About the MIPI DSI command mode to be more specific:
> We add special handling for the MIPI_DCS_WRITE_MEMORY_CONTINUE, MIPI_DCS_SET_COLUMN_ADDRESS, and other needed commands within the dsi_mcux_2l driver, similar to how we did for smartDMA (we can reuse the MCUX_DSI_2L_FB_DATA flag here), which configures the eLCDIF to send the frame buffer. In all other commands, we just use the DSI MCUX directly.

Only MIPI_DCS_WRITE_MEMORY_START and MIPI_DCS_WRITE_MEMORY_CONTINUE these 2 DCS command are supported when LCDIF is used within dsi_mcux_2l, all other commands, we just use the DSI MCUX directly.

Also I'm a little confused:
> I don't think we should call the mipi DSI driver here- the reason is because if we enable the MIPI DSI driver with the dsi_mcux_2l driver, 

Do you mean enable the MCUxpresso SDK DSI module inside the dsi_mcux_2l other than dsi_mcux_attach function? Actually the SDK DSI module is only enabled in dsi_mcux_attach, I did not change this part, there is no other driver managing the SDK DSI IP other than dsi_mcux_2l(dsi_mcux_attach ). For G1120B0MIPI the flow is:
boot:
mcux_dcnano_lcdif_dbi_init(Init LCDIF IP)
mcux_mipi_dsi_init
rm67162_init: mipi_dsi_attach(init DSI IP)->init G1120B0MIPI panel
main:
frame update: use DSI directly to set collumn/page address, use DSI with LCDIF(by using mcux_dcnano_lcdif_dbi driver) to send frame data

danieldegrasse replied:
> Only MIPI_DCS_WRITE_MEMORY_START and MIPI_DCS_WRITE_MEMORY_CONTINUE these 2 DCS command are supported when LCDIF is used within dsi_mcux_2l, all other commands, we just use the DSI MCUX directly.

Great- then we should just implement special handling for these commands

> Do you mean enable the MCUxpresso SDK DSI module inside the dsi_mcux_2l other than dsi_mcux_attach function? Actually the SDK DSI module is only enabled in dsi_mcux_attach, I did not change this part, there is no other driver managing the SDK DSI IP other than dsi_mcux_2l(dsi_mcux_attach ). For G1120B0MIPI the flow is:
boot:
mcux_dcnano_lcdif_dbi_init(Init LCDIF IP)
mcux_mipi_dsi_init
rm67162_init: mipi_dsi_attach(init DSI IP)->init G1120B0MIPI panel
main:
frame update: use DSI directly to set collumn/page address, use DSI with LCDIF(by using mcux_dcnano_lcdif_dbi driver) to send frame data

Thanks for the clarification here- I see what you mean (that the display write only uses the LCDIF driver). I think this is ok, given that- but I do think we should include a comment here making it clear what is being done (namely that the MIPI DBI driver is using the LCDIF to accelerate display writes)

I appreciate you clarifying how the IP is being used here- so let's keep the call to `mipi_dbi_write_display`, and include some form of comment clarifying that the eLCDIF is used by that API. Other than that I think we just need some clean up regarding how the special commands are handled here.

danieldegrasse replied:
One thing I am still unclear on though- for MIPI DSI mode, we can use the `LCDIF_DbiSendCommand` API calls to send MIPI-DSI commands? My understanding was that those calls would send DBI commands over an 6800 or 8080 bus.

KATE-WANG-NXP replied:
> One thing I am still unclear on though- for MIPI DSI mode, we can use the LCDIF_DbiSendCommand API calls to send MIPI-DSI commands? 

The API is only called to handle MIPI_DCS_WRITE_MEMORY_START or MIPI_DCS_WRITE_MEMORY_CONTINUE here, for DCS long write the command and data are all handled by LCDIF altogether. Other DCS commands are all ineffective for LCDIF used in MIPI-DSI command mode.

> My understanding was that those calls would send DBI commands over an 6800 or 8080 bus.

Yes LCDIF_DbiSendCommand are used in 6800 or 8080 bus, except only for DCS long write is treated specially.

kartben commented:
> @kartben I merged the HAL and rebased this PR for @KATE-WANG-NXP

compliance check issues in board.c

KATE-WANG-NXP commented on include/zephyr/drivers/display.h:
```diff
@@ -129,6 +129,8 @@ struct display_buffer_descriptor {
 	/** Indicates that this is not the last write buffer of the frame */
 	bool frame_incomplete;
+	/** Indicates that this the first write buffer of the frame */
+	bool first_write;
```
Because if LCDIF DBI is used in DSI_MCUX_2L it need to know whether to use write memory start(0x2C) or continue (0x3C). frame_incomplete is used to indicate if it is the last write not the first?

danieldegrasse replied:
The idea of `frame_incomplete` is the same as `first_frame`- if a given descriptor has `frame_incomplete` set to false, you know the next write will be a new frame (and should use `write_memory_start`). That being said, the display driver should be determining when to use write_memory_start versus continue, right? Or does the DSI driver need to handle this for some reason?

See here for an example of using `frame_incomplete` to determine if a write will be the first one: https://github.com/zephyrproject-rtos/zephyr/blob/ee8266797c176b67867c1012d441141885bc9e4c/drivers/mipi_dbi/mipi_dbi_nxp_lcdic.c#L444

danieldegrasse commented on drivers/mipi_dbi/mipi_dbi_nxp_dcnano_lcdif.c:
```diff
@@ -0,0 +1,280 @@
+		LCDIF_SetFrameBufferStride(config->base, 0, bytes_per_pixel * desc->pitch);
+	}
+
+	LCDIF_SetFrameBufferPosition(config->base, 0U, 0U, 0U, desc->width,
```
This setup looks like it will only support full frames, not partial frames. Is that by design?

KATE-WANG-NXP replied:
Adding more comment for clarification.

danieldegrasse commented on drivers/mipi_dbi/mipi_dbi_nxp_dcnano_lcdif.c:
```diff
@@ -0,0 +1,280 @@
+	LCDIF_SetFrameBufferAddr(config->base, 0, (uint32_t)framebuf);
+
+	/* Send command: Memory write start or continue. */
+	if (desc->frame_incomplete) {
```
We can't rely on `frame_incomplete` always being set like this- it is false by default, and only set to true by some drivers. Take a look at the initial PR adding this, the intention was more to indicate when a new frame would be starting: https://github.com/zephyrproject-rtos/zephyr/pull/81250

The idea of the `write_display` API is that the MIPI DBI display driver will send the `DCS_WRITE_MEMORY` command as a separate call to the `mipi_dbi_command_write` API, then the `write_display` API will be used to send the graphical data *only*. So we shouldn't need this here- the MIPI DSI driver should be able to send the DSI command itself

KATE-WANG-NXP replied:
I see! Update accordingly.

danieldegrasse commented on boards/shields/rk055hdmipi4ma0/boards/mimxrt700_evk_mimxrt798s_cm33_cpu0.overlay:
```diff
@@ -0,0 +1,20 @@
+};
+
+&zephyr_mipi_dsi {
+	/* Use lower DPHY clock since RT700 uses AUDIO pll pfd2 as aource,
```
```suggestion
	/* Use lower DPHY clock since RT700 uses AUDIO pll pfd2 as source,
```
Nit

danieldegrasse commented on drivers/mipi_dsi/dsi_mcux_2l.c:
```diff
@@ -560,14 +581,107 @@ static ssize_t dsi_mcux_transfer(const struct device *dev, uint8_t channel,
 	}
 
-	status = DSI_TransferBlocking(config->base, &dsi_xfer);
+#if CONFIG_MIPI_DSI_MCUX_NXP_DCNANO_LCDIF
```
Would it make sense to define two different functions for `dsi_mcux_transfer`? It seems like we are effectively replacing the entire function when `CONFIG_MIPI_DSI_MCUX_NXP_DCNANO_LCDIF` is set- so maybe we define two versions of the function (one for the RT500, and one for the RT700)?

KATE-WANG-NXP replied:
The lcdif only helps with long write, for other commands it still requires mipi-dsi, if using 2 functions there will be too many shared codes, so maybe create another function for special handling of lcdif part?

dleach02 commented:
@KATE-WANG-NXP @danieldegrasse Are we converging on being able to approve and merge?

Also, @KATE-WANG-NXP please rebase to resolve the conflict.

dleach02 commented:
@KATE-WANG-NXP please address the merge conflicts

mmahadevan108 commented on boards/shields/rk055hdmipi4ma0/boards/mimxrt700_evk_mimxrt798s_cm33_cpu0.overlay:
```diff
@@ -0,0 +1,20 @@
+ */
+
+&zephyr_lcdif {
+	version = "DC8000";
```
Please define the `version` in the SoC `dts` file as this will not change.

KATE-WANG-NXP replied:
Hi, by default the zephyr_lcdif of RT700 uses compatible "nxp,dcnano-lcdif-dbi" while 'version' belongs to the compatible  "nxp,dcnano-lcdif", so actually I shall define the default compatible "nxp,dcnano-lcdif-dbi" rather than "nxp,dcnano-lcdif" in the SoC dts file, and keep this version inside this overlay, thanks for the remind.

danieldegrasse replied:
@decsny are you ok with having two compatibles here for different drivers? Or should we have one compatible and decide which driver to use based on Kconfig?

mmahadevan108 commented on boards/shields/rk055hdmipi4ma0/boards/mimxrt700_evk_mimxrt798s_cm33_cpu0.overlay:
```diff
@@ -0,0 +1,20 @@
+
+&zephyr_lcdif {
+	version = "DC8000";
+		display-timings {
```
Please check the indentation of this line.

mmahadevan108 commented on boards/shields/g1120b0mipi/boards/mimxrt700_evk_mimxrt798s_cm33_cpu0.overlay:
```diff
@@ -0,0 +1,21 @@
+
+&zephyr_lcdif {
+	status = "okay";
+	compatible = "nxp,dcnano-lcdif-dbi";
```
Where is this binding file defined?

KATE-WANG-NXP replied:
dts/bindings/mipi-dbi/nxp,mipi-dbi-dcnano-lcdif.yaml

KATE-WANG-NXP replied:
Yes thanks for the remind

KATE-WANG-NXP commented on boards/nxp/mimxrt700_evk/mimxrt700_evk_mimxrt798s_cm33_cpu0.dts:
```diff
@@ -115,6 +151,14 @@
+&gpio3 {
+	status = "okay";
+};
+
```
HI, the TE input pin uses GPIO3:
https://github.com/zephyrproject-rtos/zephyr/blob/558c1f3873c0de3b20d318ffea2fe7bc8fcab480/boards/nxp/mimxrt700_evk/mimxrt700_evk_mimxrt798s_cm33_cpu0.dts#L75

KATE-WANG-NXP committed:
dts: arm: nxp: update RT7xx dts files for LCDIF and MIPI-DSI

Add configuration for LCDIF and MIPI-DSI.

Signed-off-by: Kate Wang <yumeng.wang@nxp.com>

KATE-WANG-NXP commented:
Hi @Iuliana Prodan, twister build with errors of 
![image](https://github.com/user-attachments/assets/3df0dffe-6a63-4f14-93a4-d599b23aa793)
https://github.com/zephyrproject-rtos/zephyr/actions/runs/14193303154/job/39762960144?pr=82635
May you please help have a look?
Also @hakehuang

zejiang0jason commented:
> Hi @iuliana-prodan , twister build with errors of ![image](https://private-user-images.githubusercontent.com/82034034/429299453-3df0dffe-6a63-4f14-93a4-d599b23aa793.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM1NjM5MjcsIm5iZiI6MTc0MzU2MzYyNywicGF0aCI6Ii84MjAzNDAzNC80MjkyOTk0NTMtM2RmMGRmZmUtNmE2My00ZjE0LTkzYTQtZDU5OWIyM2FhNzkzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDAyVDAzMTM0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIwZjdjYzVjNDYxNzkxMGVjMDljZjg3MTEzNGMzNThhYjIxODAwYTc3NzdjOGM3ZTY0MzM0YzU4OTEyY2Q5NTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.1XNfVoNIenFOBM5ncsmXdVFX0gNA2LttBbbvVNv4dKQ) https://github.com/zephyrproject-rtos/zephyr/actions/runs/14193303154/job/39762960144?pr=82635 May you please help have a look? Also @hakehuang

Hi @KATE-WANG-NXP , this is due to recent hal code upgrade, we are working on it now.

KATE-WANG-NXP committed:
samples: drivers: display: Enable example for g1120b0mipi on mimxrt798s

Enable display example for g1120b0mipi shield on mimxrt798s board and
update board document.

Signed-off-by: Kate Wang <yumeng.wang@nxp.com>

iuliana-prodan commented:
> Hi @iuliana-prodan , twister build with errors of ![image](https://private-user-images.githubusercontent.com/82034034/429299453-3df0dffe-6a63-4f14-93a4-d599b23aa793.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM1ODA5OTgsIm5iZiI6MTc0MzU4MDY5OCwicGF0aCI6Ii84MjAzNDAzNC80MjkyOTk0NTMtM2RmMGRmZmUtNmE2My00ZjE0LTkzYTQtZDU5OWIyM2FhNzkzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDAyVDA3NTgxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJmMjZjM2Y1MzNhYjJjNjQ5YWUyYWQwYjM5YmNmMWYzNmUyYzRhZDA5MTBjMDc4MjgyYjk0NDJhNjc4YWM5NWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.h-ug3QmGdA6EwcFbppFMOY9P23q48Z-Mas9e-JfTLYg) https://github.com/zephyrproject-rtos/zephyr/actions/runs/14193303154/job/39762960144?pr=82635 May you please help have a look? Also @hakehuang

@ZhaoxiangJin has a fix for this: https://github.com/zephyrproject-rtos/hal_nxp/pull/525

hakehuang commented:
> > Hi @iuliana-prodan , twister build with errors of ![image](https://private-user-images.githubusercontent.com/82034034/429299453-3df0dffe-6a63-4f14-93a4-d599b23aa793.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM1ODA5OTgsIm5iZiI6MTc0MzU4MDY5OCwicGF0aCI6Ii84MjAzNDAzNC80MjkyOTk0NTMtM2RmMGRmZmUtNmE2My00ZjE0LTkzYTQtZDU5OWIyM2FhNzkzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDAyVDA3NTgxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJmMjZjM2Y1MzNhYjJjNjQ5YWUyYWQwYjM5YmNmMWYzNmUyYzRhZDA5MTBjMDc4MjgyYjk0NDJhNjc4YWM5NWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.h-ug3QmGdA6EwcFbppFMOY9P23q48Z-Mas9e-JfTLYg) https://github.com/zephyrproject-rtos/zephyr/actions/runs/14193303154/job/39762960144?pr=82635 May you please help have a look? Also @hakehuang
> 
> @ZhaoxiangJin has a fix for this: [zephyrproject-rtos/hal_nxp#525](https://github.com/zephyrproject-rtos/hal_nxp/pull/525)

I need you help to self-test this, as I have no LCD board for RT700.

danieldegrasse commented on dts/bindings/mipi-dbi/nxp,mipi-dbi-dcnano-lcdif.yaml:
```diff
@@ -0,0 +1,136 @@
+include: [base.yaml]
+
+properties:
+  mipi-mode:
```
The `mipi-mode` should be set by the display controller on the MIPI DBI bus, not the MIPI DBI controller- so this property should not be present here, instead the MIPI DBI driver should select the mode based on the `mode` field in the `mipi_dbi_config` structure

KATE-WANG-NXP replied:
Yes, also the DBI spec specify not only the bus type A/B/C and data width but also the color coding like RGB332 or RGB888. So it seems the MIPI_DBI_MODE_xxx defined in https://github.com/nxp-upstream/zephyr/blob/95d41f85f721fbd838713533e850edcbd450eebf/include/zephyr/dt-bindings/mipi_dbi/mipi_dbi.h#L81 is not sufficient.(I'm guessing that the previous driver do not have such need for this configuration so it is not included?) I'll add more items in include/zephyr/dt-bindings/mipi_dbi/mipi_dbi.h for the color coding and remove this mipi-mode and also remove the color-coding defined here:
https://github.com/zephyrproject-rtos/zephyr/commit/aa47658b352f9c9bf1ce5d07cdcf85e0a2a93550#diff-e6135917010f7d34fcb7fcb792865e5d09f6cb006b794f85c44b942fe3c57c91R112
And they will be configured by display controller

danieldegrasse replied:
As far as I understand, MIPI-DBI is only defined with 3 modes, and no mention of color coding. Is there a different specification you are looking at? The LCDIF can set the color coding, but the display controller on the bus should set the interface type. In most cases the LCDIF should not need to be aware of the color coding, right? Unless some kind of swizzling is occurring when sending pixel data

KATE-WANG-NXP replied:
Hi @danieldegrasse , this is the MIPI_DBI_Specification_v2.pdf https://members.mipi.org/wg/All-Members/document/9203

danieldegrasse replied:
Ah ok- this is what I get for not having access to the latest MIPI-DBI spec. Thank you for catching this. It looks like you extended the MIPI mode definitions, which is exactly what I think we want here

danieldegrasse commented on dts/bindings/mipi-dbi/nxp,mipi-dbi-dcnano-lcdif.yaml:
```diff
@@ -0,0 +1,136 @@
+      - "D16-RGB888-Option1" # For type A or B. 2 pixels sent in 3 cycles.
+      - "D16-RGB888-Option2" # For type A or B. 1 pixel sent in 2 cycles.
+
+  mipi-dsi:
```
Where is this used?

KATE-WANG-NXP replied:
Sorry, deleted.

danieldegrasse commented on drivers/mipi_dsi/dsi_mcux_2l.c:
```diff
@@ -560,14 +581,107 @@ static ssize_t dsi_mcux_transfer(const struct device *dev, uint8_t channel,
 	}
 
-	status = DSI_TransferBlocking(config->base, &dsi_xfer);
+#ifdef CONFIG_MIPI_DSI_MCUX_NXP_DCNANO_LCDIF
```
Could you split this into a separate function? That way the code flow within the transfer function is a bit easier to follow- we can just call this function if `CONFIG_MIPI_DSI_MCUX_NXP_DCNANO_LCDIF` is set

KATE-WANG-NXP replied:
Yes , updated.

danieldegrasse reviewed (commented):
Generally ok from my side- only specific issue I see is that we should not be setting the `mipi-mode` in the DBI controller binding, we should be configuring it based on what the attached DBI display requests. It would also be great to see a DTS definition for something like the LCD-PAR display, since that supports 6800/8080 mode- if that isn't possible though no worries!

ZhaoxiangJin commented:
there are some CI build failures, i submitted PR https://github.com/zephyrproject-rtos/zephyr/pull/88305 to fix it.

mmahadevan108 commented on dts/arm/nxp/nxp_rt7xx_cm33_cpu0.dtsi:
```diff
@@ -1077,6 +1078,27 @@
 	};
+
+	lcdif: lcdif@480000 {
+		compatible = "nxp,mipi-dbi-dcnano-lcdif";
```
Can we define the `version` property here since it will not change.

KATE-WANG-NXP replied:
Hi, the lcdif has 2 drivers with 2 compatible, the version property belongs to the other compatible ""nxp,dcnano-lcdif

mmahadevan108 commented on boards/nxp/mimxrt700_evk/doc/index.rst:
```diff
@@ -218,3 +218,24 @@ should see the following message in the terminal:
+   :shield: g1120b0mipi
+   :zephyr-app: samples/drivers/display
+   :goals: build
+   :compact:
```
Do we need a section for `rk055hdmipi4ma0` display?

KATE-WANG-NXP commented:
> there are some CI build failures, i submitted PR #88305 to fix it.

Please resolve the conflicts.

KATE-WANG-NXP committed:
boards: nxp: add configuration for MIPI-DSI and LCDIF on mimxrt700_evk

1.Add clock configuration for DCNano LCDIF video and
command mode.
2.Update board dts file

Signed-off-by: Kate Wang <yumeng.wang@nxp.com>

KATE-WANG-NXP committed:
boards: shields: Support panel g1120b0mipi and rk055hdmipi4ma0 on RT700

Add board specific overlay and configuration for display panel
g1120b0mipi and rk055hdmipi4ma0.

Signed-off-by: Kate Wang <yumeng.wang@nxp.com>

KATE-WANG-NXP committed:
drivers: mipi_dsi: dsi_mcux_2l: Use NXP DCNano DBI driver for memory write

There is no smartdma on RT700, so to perform DCS memory write the CPU has
to write APB buffer word by word, which is too slow for most applications.
But the DCNano in DBI mode can be used to interface with the MIPI-DSI on
RT700, and send data to MIPI-DSI to transfer, once it is properly
configured, which solves the issue.
First added new parameter first_write in display_buffer_descriptor to let
NXP DCNano DBI driver know to use MIPI_DCS_WRITE_MEMORY_START or
MIPI_DCS_WRITE_MEMORY_CONTINUE.
Second updated the MCUX MIPI-DSI driver to support using the NXP DCNano DBI
driver for memory write.

Signed-off-by: Kate Wang <yumeng.wang@nxp.com>

KATE-WANG-NXP committed:
soc: nxp: imxrt: imxrt7xx: update rt7xx soc files

Add functions to configure MIPI_DSI power and clock when
MIPI_DSI is enabled.

Signed-off-by: Kate Wang <yumeng.wang@nxp.com>

danieldegrasse commented on include/zephyr/dt-bindings/mipi_dbi/mipi_dbi.h:
Can you pull this change into a separate commit since it touches generic code? The commit message should make it clear why we are adding these (IE to comply with the DBI spec) for anyone looking in the future

KATE-WANG-NXP replied:
Yes, updated.

danieldegrasse commented:
Once CI error is cleared and the `mipi_dbi.h` change is moved to a separate commit, I'm good for this PR to move forwards. Apologies for all the delays here @KATE-WANG-NXP.

KATE-WANG-NXP committed:
drivers: mipi_dbi: introduce NXP DCnano driver

Introduce NXP NCNano driver using MIPI DBI class. This peripheral
supports 8080 and 6800 mode. The driver also supports used with
nxp,mipi_dsi_2l driver, for the panel with DPHY bus, such as g1120b0mipi.

Signed-off-by: Kate Wang <yumeng.wang@nxp.com>

KATE-WANG-NXP committed:
drivers: mipi_dbi: Add defines for color coding

Add defines for color coding of DBI Type A(Motorola 6800)
and B(Intel 8080) bus. This is to comply with the MIPI
Alliance Standard for Display Bus Interface v2.0.

Signed-off-by: Kate Wang <yumeng.wang@nxp.com>

danieldegrasse reviewed (approved):
CI issues still need resolution, but changes themselves are ok with me

KATE-WANG-NXP commented:
Blocked by https://github.com/zephyrproject-rtos/zephyr/pull/88305, waiting for merge.

KATE-WANG-NXP commented:
Hi @dleach02 , I think this PR is ready to merge
